python/semantic_kernel/connectors/ai/function_choice_behavior.py

```diff
@@ -0,0 +1,244 @@
+    from semantic_kernel.kernel import Kernel
+
+DEFAULT_MAX_AUTO_INVOKE_ATTEMPTS = 5
+
+logger = logging.getLogger(__name__)
+
+
+@experimental_class
+class FunctionChoiceType(Enum):
+    """The type of function choice behavior."""
+
+    AUTO = "auto"
+    NONE = "none"
+    REQUIRED = "required"
+
+
+@experimental_function
+def _check_for_missing_functions(function_names: list[str], kernel_function_metadata: set) -> None:
+    """Check for missing functions in the kernel function metadata."""
+    try:
+        qualified_names = {metadata.fully_qualified_name for metadata in kernel_function_metadata}
+        missing_functions = [name for name in function_names if name not in qualified_names]
+    except Exception:
+        logger.exception("Error checking for missing functions.")
+        return
+
+    if missing_functions:
+        raise ServiceInvalidExecutionSettingsError(
+            f"Unable to find the required function(s) `{', '.join(missing_functions)}` "
+            f"in the list of available kernel functions: {', '.join(qualified_names)}."
+        )
+
+
+@experimental_class
+@dataclass
+class FunctionCallChoiceConfiguration:
+    available_functions: list["KernelFunctionMetadata"] | None = None
+
+
+@experimental_class
+class FunctionChoiceBehavior(KernelBaseModel):
+    """Class that controls function choice behavior.
+
+    Attributes:
+        enable_kernel_functions: Enable kernel functions.
+        max_auto_invoke_attempts: The maximum number of auto invoke attempts.
+        filters: Filters for the function choice behavior.
+        function_fully_qualified_names: The fully qualified names of the functions.
+        type: The type of function choice behavior.
+
+    Properties:
+        auto_invoke_kernel_functions: Check if the kernel functions should be auto-invoked.
+            Determined as max_auto_invoke_attempts > 0.
+
+    Methods:
+        configure: Configures the settings for the function call behavior,
+            the default version in this class, does nothing, use subclasses for different behaviors.
+
+    Class methods:
+        Auto: Returns FunctionChoiceAuto class with auto_invoke enabled, and the desired functions
+            based on either the specified filters or the full qualified names.
+        NoneInvoke: Returns FunctionChoiceNone class with auto_invoke disabled, and the desired functions
+            based on either the specified filters or the full qualified names.
+        Required: Returns FunctionChoiceRequired class with auto_invoke enabled, and the desired functions
+            based on either the specified filters or the full qualified names.
+    """
+
+    enable_kernel_functions: bool = True
+    maximum_auto_invoke_attempts: int = DEFAULT_MAX_AUTO_INVOKE_ATTEMPTS
+    filters: (
+        dict[Literal["excluded_plugins", "included_plugins", "excluded_functions", "included_functions"], list[str]]
```

Comment:
Just a question: how do the `included_xxx`s work? Do they include only the ones in the list and exclude all others? Are they mutually exclusive to the `excluded_xxx` lists?

Reply:
That is correct. They are mutually exclusive and an error is thrown if both `included_*` and `excluded_*` are provided. Otherwise if one specifies `{"included_plugins": ["math", "time"], "included_functions": ["math-Add", "time-time"]}`, the `included_functions` supersede the `included_plugins`.

Reply:
Is this documented and discoverable somewhere?